lib/smf/smf.c

```diff
@@ -199,98 +242,156 @@ void smf_set_initial(struct smf_ctx *ctx, const struct smf_state *init_state)
 		init_state = init_state->initial;
 	}
 #endif
+
 	internal->exit = false;
 	internal->terminate = false;
 	ctx->current = init_state;
 	ctx->previous = NULL;
 	ctx->terminate_val = 0;
 
-	if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT)) {
-		internal->new_state = false;
+	internal->entry = true;
+
+#ifdef CONFIG_SMF_ANCESTOR_SUPPORT
```

Comment:
I think if you eliminate the #ifdef around the static functions above that have the `__unused` attribute, you can switch back to using `if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT))`, which is preferred.

The compiler should properly exclude the code inside the `if IS_ENABLED` block as well as the unused functions.

Reply:
Just wondering why `if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT))` is preferred. I'm guessing the compiler optimizes it out, or else all the functions marked unused would be included?

Reply:
Is there a historical reason for preferring `if (IS_ENABLED(CONFIG_SMF_ANCESTOR_SUPPORT))` to `#if`?

Reply:
https://www.kernel.org/doc/html/v4.10/process/coding-style.html#conditional-compilation